Feature/bjs2 66218 add user avatar

marsel-mkh opened the pull request.

Commit:
BJS2-66218 firs stage

Commit:
BJS2-66218 firs stage fix yaml

Commit:
BJS2-66218 упорядочены файлы по папкам
выполнена таска "Добавление аватара пользователя" с учетом маштабируемости, правильного возврата данных на фронт в методе get.

marsel-mkh commented:
Скажу сразу: возможно много всяких косяков. Например разные типы конфигураций в aws и avatar, но это могу оправдать, тем что в aws хотелось скрыть логин и пароль:)

Commit:
BJS2-66218 fix any small problems,delete spaces

Commit:
BJS2-66218 added test

safon4ik10 commented on src/main/java/school/faang/user_service/controller/avatar/AvatarController.java:
```diff
@@ -0,0 +1,73 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/users/{userId}/avatar")
```
Мне кажется это не совсем корректно, делать такой путь для самого контроллера

marsel-mkh replied:
Я думаю с учетом того что это не основной функционал, а больше дополнительный и вынесен в отдельный контроллер, то норм. Тем более зачем дублировать "/users/{userId}/avatar" в каждом методе, если можно вынести?

Zhltsk-V commented on src/main/java/school/faang/user_service/config/avatar/AvatarConfig.java:
```diff
@@ -0,0 +1,21 @@
+@Data
+@RequiredArgsConstructor
+@ConfigurationProperties(prefix = "avatar")
+public class AvatarConfig {
```
Я бы добавил валидацию полей аннотациями (@Min, @NotNull и т.д.)

Zhltsk-V commented on src/main/java/school/faang/user_service/config/aws/AwsConfig.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Bean
+    public AmazonS3 amazonS3() {
+        AWSCredentials credentials = new BasicAWSCredentials(accessKey, secretKey);
```
Нет обработки ошибок при создании клиента, вот пример:       
        
AWSCredentials credentials = new BasicAWSCredentials(
   Objects.requireNonNull(accessKey, "Minio access key must not be null"),
   Objects.requireNonNull(secretKey, "Minio secret key must not be null")
);

Zhltsk-V commented on src/main/java/school/faang/user_service/s3/S3Resource.java:
```diff
@@ -0,0 +1,96 @@
+ *
+ */
+@RequiredArgsConstructor
+public class S3Resource extends AbstractResource implements AutoCloseable {
```
Нет проверки s3Object на null

Zhltsk-V commented on src/main/java/school/faang/user_service/s3/S3Service.java:
```diff
@@ -0,0 +1,91 @@
+     * @throws FileException если произошла ошибка при загрузке
+     *
+     */
+    public String uploadFile(MultipartFile file, String folder) {
```
Я бы сюда добавил аннотацию @Transactional(propagation = Propagation.REQUIRED), возможно она нужна и в других методах =)

Zhltsk-V replied:
Думаю стоит сюда добавить валидацию входных параметров (можно сделать для этого отдельный метод)

marsel-mkh replied:
S3 не поддерживает транзакции в классическом понимании. Операция putObject выполняется атомарно: либо файл загружается полностью, либо нет.

Zhltsk-V commented on src/main/java/school/faang/user_service/s3/S3Service.java:
```diff
@@ -0,0 +1,91 @@
+        ObjectMetadata objectMetadata = new ObjectMetadata();
+        objectMetadata.setContentLength(fileSize);
+        objectMetadata.setContentType(file.getContentType());
+        String key = String.format("%s/%s", folder, file.getOriginalFilename());
```
Давай вынесем эту логику получения ключа в отдельный метод

Zhltsk-V commented on src/main/java/school/faang/user_service/s3/S3Service.java:
```diff
@@ -0,0 +1,91 @@
+            throw new FileException(e.getMessage());
+        }
+    }
+}
```
Подумай над добавление логгирования значимых операций

Zhltsk-V commented on src/main/java/school/faang/user_service/service/avatar/AvatarService.java:
```diff
@@ -0,0 +1,149 @@
+@Service
+@RequiredArgsConstructor
+public class AvatarService {
+    private final String SMALL_AVATAR_SIZE = "small";
```
Константы SMALL_AVATAR_SIZE и LARGE_AVATAR_SIZE можно вынести в конфиг

Zhltsk-V commented on src/main/java/school/faang/user_service/service/avatar/AvatarService.java:
```diff
@@ -0,0 +1,149 @@
+    @Transactional
+    public void deleteUserAvatar(long userId) {
+        User user = userService.getUser(userId);
+        UserProfilePic userProfilePic = user.getUserProfilePic();
```
Можно проверить на null

Zhltsk-V commented on src/main/java/school/faang/user_service/service/avatar/AvatarService.java:
```diff
@@ -0,0 +1,149 @@
+     * @throws IllegalArgumentException если файл аватара не проходит валидацию
+     */
+    @Transactional
+    public void addUserAvatar(long userId, MultipartFile avatar) {
```
Думаю имеет смысл добавить валидацию поля avatar на null

Zhltsk-V commented on src/main/java/school/faang/user_service/service/avatar/AvatarService.java:
```diff
@@ -0,0 +1,149 @@
+ */
+@Service
+@RequiredArgsConstructor
+public class AvatarService {
```
Класс можно пометить аннотацией @Transactional

Zhltsk-V commented on src/main/java/school/faang/user_service/service/avatar/ImageResize.java:
```diff
@@ -0,0 +1,57 @@
+     * @apiNote Всегда сохраняет результат в формате JPG независимо от исходного формата
+     * @implNote Использует алгоритм масштабирования Scalr.Method.QUALITY для наилучшего качества
+     */
+    public MultipartFile resizeImage(MultipartFile file, int maxSize) {
```
Отсутствует валидация входных параметров и нет ограничения на минимальный размер изображения

Zhltsk-V commented on src/main/java/school/faang/user_service/validation/avatar/AvatarValidator.java:
```diff
@@ -0,0 +1,36 @@
+     * @param avatar файл аватара для проверки (не может быть null)
+     * @throws DataValidationException если размер файла превышает допустимый*
+     */
+    public void checkAvatarSize(MultipartFile avatar) {
```
Сюда можно добавить проверку на формат файла

Zhltsk-V commented on src/main/resources/application.yaml:
```diff
@@ -34,4 +34,20 @@ project-service:
+  max-size-bytes: 5242880    # 5MB (макс. размер файла)
+  folder-template: "users/{username}/avatars"
+  sizes:
+    large: 1080
```
Поля small и large можно использовать в сервисе, на данный момент ты их не используешь. Вместо это ты создаешь константы SMALL_AVATAR_SIZE и LARGE_AVATAR_SIZE и используешь их.

Zhltsk-V commented on src/main/java/school/faang/user_service/s3/S3Resource.java:
```diff
@@ -0,0 +1,96 @@
+     *
+     */
+    @Override
+    public void close() throws IOException {
```
Ты сначала проверяется getInputStream(), а потом s3Object мне кажется было бы логичнее наоборот

marsel-mkh replied:
Ресурсы закрываются именно в таком порядке, сначала стрим данных, а затем обьект. Если закрыть обьект, то как потом закрыть инпут стрим?

Zhltsk-V commented on src/main/java/school/faang/user_service/s3/S3Service.java:
```diff
@@ -0,0 +1,91 @@
+     * @param fileId ключ (путь) к файлу в S3 (например, "users/123/avatars/profile.jpg")
+     * @return InputStream с содержимым файла
+     */
+    public Resource downloadFile(String fileId) {
```
В downloadFile нет проверки на существование файла перед скачиванием

Zhltsk-V commented on src/test/java/school/faang/user_service/controller/AvatarControllerTest.java:
```diff
@@ -0,0 +1,78 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@WebMvcTest(controllers = AvatarController.class)
+public class AvatarControllerTest {
```
Добавб тесты для случаев, когда сервис выбрасывает исключения

Zhltsk-V commented on src/test/java/school/faang/user_service/service/S3ServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class S3ServiceTest {
```
Добавь тест для случая, когда файл не существует при скачивании

Zhltsk-V commented on src/test/java/school/faang/user_service/validation/goal/AvatarValidatorTest.java:
```diff
@@ -0,0 +1,78 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class AvatarValidatorTest {
```
Можно добавить проверку пустого файла и обработку null файла

Commit:
BJS2-66218 added validation

Commit:
BJS2-66218 fix bugs

RinatBeybutov commented on src/main/java/school/faang/user_service/controller/avatar/AvatarController.java:
```diff
@@ -0,0 +1,73 @@
+            summary = "Upload user avatar",
+            description = "Upload or update a user's profile picture"
+    )
+    @PostMapping("/add")
```
давай уберем здесь эти add delete и тд в эндпоинте. Тк у тебя post delete и get позволяют выбрать метод контроллера и нигде нет дублирования

RinatBeybutov commented on src/main/java/school/faang/user_service/service/avatar/AvatarService.java:
```diff
@@ -0,0 +1,151 @@
+     * @return путь к папке пользователя
+     */
+    private String buildUserFolder(User user) {
+        return avatarConfig.getFolderTemplate().replace("{username}", user.getUsername());
```
давай на 2 строки разнесем, длинная строчка получилась

RinatBeybutov commented on src/main/java/school/faang/user_service/service/avatar/ImageResize.java:
```diff
@@ -0,0 +1,58 @@
+     */
+    public MultipartFile resizeImage(@NotNull MultipartFile file, int maxSize) {
+        ByteArrayOutputStream outputStream;
+        try {
```
давай outputStream внесем в try и return тоже в try внесем

RinatBeybutov replied:
https://www.baeldung.com/java-try-with-resources

Commit:
BJS2-6621
FIX:  delete  mapping endpoints in AvatarController
change try-catch in ImageResize

Commit:
BJS2-6621
FIX: AvatarControllerTest fixing

Commit:
BJS2-66218
FIX: use try-with-resources in resizeImage
